declfilter/declfilter.go

```diff
@@ -3,20 +3,92 @@
 package declfilter
 
 import (
+	"regexp"
+	"strings"
+
 	"github.com/pelletier/go-toml/v2"
 )
 
-// FilterConfig defines the format of the toml files
-//
-// TODO: currently the only way to "wildcard" a config is to not have a toml
-// file. Should have something more flexible like glob patterns.
-type FilterConfig struct {
+// TODO: describe semantic interpretation for stringSet.
+
+type setOpType int
+
+const (
+	setUnion setOpType = iota
+	setSubtract
+)
+
+type setOp struct {
+	t setOpType
+	r *regexp.Regexp
+}
+
+// A string set described by a sequence of glob patterns.
+type stringSet []setOp
+
+func (ss stringSet) contains(s string) bool {
+	b := false
+	for _, op := range ss {
+		if op.r.MatchString(s) {
+			switch op.t {
+			case setUnion:
+				b = true
+			case setSubtract:
+				b = false
+			}
+		}
+	}
+	return b
+}
+
+func newOp(pat string) setOp {
+	var s setOp
+	pattern, negated := strings.CutPrefix(pat, "!")
+	if negated {
+		s.t = setSubtract
+	} else {
+		s.t = setUnion
+	}
+
+	patternParts := strings.Split(pattern, "*")
+	for i := range patternParts {
+		patternParts[i] = regexp.QuoteMeta(patternParts[i])
+	}
+	var err error
+	s.r, err = regexp.Compile("^" + strings.Join(patternParts, ".*") + "$")
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+func sliceMap[Slice ~[]A, A any, B any](s Slice, f func(A) B) []B {
+	result := make([]B, len(s))
+	for i, v := range s {
+		result[i] = f(v)
+	}
+	return result
+}
+
+func newStringSet(s []string) stringSet {
+	return sliceMap(s, newOp)
+}
+
+// filterConfig defines the format of the toml files
+type filterConfig struct {
```

Comment:
I made this public on the grounds that it's the documentation of the toml file format, which is the most important public API of this package.